### PR TITLE
mrs: realloc() compat fix

### PR DESCRIPTION
**Pull request by brooksdavis**
Fix realloc() to return non-NULL with a size of 0 to fix sort and others (git grep?). Add test of this property.

### CODE REVIEW

**Comment by brooksdavis** [comment]
My initial push accidentally included #1887. Should merge it first or rebase.

**Review by markjdb — APPROVED** [review]
I hadn't looked at mrs before, but this looks sane to me.

File: bin/cheribsdtest/cheribsdtest_malloc.c [hunk 1]
```diff
@@ -71,3 +73,47 @@ CHERIBSDTEST(malloc_revoke_twice, "revoke twice back to back",
...
+	free(ptr);
+	CHERIBSDTEST_VERIFY((ptr = calloc(0, 0)) != NULL);
+	free(ptr);
+
+	CHERIBSDTEST_VERIFY((ptr = realloc(NULL, 0)) != NULL);
+	CHERIBSDTEST_VERIFY((ptr2 = realloc(ptr, 0)) != NULL);
+	/*
+	 * XXX: POSIX requires that: "A pointer to the allocated space
+	 * shall be returned, and the memory object pointed to by ptr
+	 * shall be freed."  Unfortunatly that's impossible to check as
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest_malloc.c:97]:
```suggestion
	 * shall be freed."  Unfortunately that's impossible to check as
```

File: lib/libc/stdlib/malloc/mrs/mrs.c [hunk 2]
```diff
@@ -1257,7 +1257,7 @@ static void *mrs_realloc(void *ptr, size_t size) {
 
 	if (size == 0) {
 		mrs_free(ptr);
-		return NULL;
+		ptr = NULL;
```
Comment (bsdjhb) [inline lib/libc/stdlib/malloc/mrs/mrs.c:1260]: I actually think we might want a different fix for this (see my last fixup in #1891).  Namely, if realloc() returns NULL, that indicates an error and that the original pointer has _not_ been free'd.  If for some reason mrs_malloc(0) returned NULL, we need to not free the old object, so my fixup commit there just removes the 'size == 0' case entirely.

**Review by bsdjhb — COMMENTED** [review]
This does not help with git grep, but it fixes sort(1) and Xorg.

**Comment by brooksdavis** [comment]
rebased on #1891

**Review by bsdjhb — APPROVED** [review]

**Pull request merged**